builder/dockerfile/builder.go

```diff
@@ -347,6 +347,26 @@ func (b *Builder) Cancel() {
 	b.cancel()
 }
 
+func (b *Builder) pullOrGetImage(name string) (builder.Image, error) {
```

Comment:
I moved this to `Builder` because it's tightly coupled, It depends on 4 fields on the Builder. It makes it easier to expose this function from `imageContexts` (as an interface with a single method)